python/tank/bootstrap/manager.py

```diff
@@ -221,49 +247,160 @@ def set_progress_callback(self, callback):
         """
         self._progress_cb = callback
 
-    def bootstrap_engine(self, engine_name, entity=None):
+    def bootstrap_engine(self,
+                         engine_name,
+                         entity=None,
+                         progress_callback=None,
+                         completed_callback=None,
+                         failed_callback=None):
         """
         Create an sgtk instance for the given project or site,
         then launch into the given engine.
 
+        When property ``async_bootstrap`` is false, the whole engine bootstrap logic
+        will be executed synchronously in the main application thread. When true,
+        an :class:`sgtk.Sgtk` instance will be bootstrapped asynchronously in a background thread,
+        followed by launching the engine synchronously in the main application thread.
+
         If entity is None, the method will bootstrap into the site
         config. This method will attempt to resolve the config according
         to business logic set in the associated resolver class and based
         on this launch a configuration. This may involve downloading new
         apps from the toolkit app store and installing files on disk.
 
+        3 callback functions can be provided.
+
+        A callback function that reports back on the bootstrap progress
+        with the following signature::
+
+            progress_callback(step_number, message, current_index, maximum_index)
+
+        where:
+        - ``step_number`` is the current progress step number,
+                          from 1 to ``ToolkitManager.MAX_PROGRESS_STEP_NUMBER``.
+        - ``message`` is the progress message to report.
+        - ``current_index`` is an optional current item number being looped over.
+        - ``maximum_index`` is an optional maximum item number being looped over.
+
+        A callback function that handles cleanup after successful completion of the bootstrap
+        with the following signature::
+
+            completed_callback(engine)
```

Comment:
the fact that the engine is handed over via the completed callback is another good reason why we should have a separate `bootstrap_engine_async` call i think.